### PR TITLE
[TF-76][Python Interop] add PythonConvertible conformance to TensorShape

### PR DESCRIPTION
**Pull request by realdoug**
This makes TensorShape conform to PythonConvertible per [TF-76](https://bugs.swift.org/browse/TF-76).

Couple small notes:
- It looks like we can essentially delegate to the implementation in [```Array```](https://github.com/apple/swift/blob/tensorflow/stdlib/public/Python/Python.swift#L998)
- I noticed that there were no tests for ```Array``` and ```Dictionary```'s existing ```PythonConvertible``` conformance so I added those as well.

### CODE REVIEW

File: stdlib/public/TensorFlow/TensorShape.swift [hunk 1]
```diff
@@ -158,3 +160,13 @@ extension TensorShape : Codable {
     self.init(dimensions)
   }
 }
+
+extension TensorShape: PythonConvertible {
+  public var pythonObject: PythonObject {
+    return self.dimensions.pythonObject
```
Comment (rxwei) [inline stdlib/public/TensorFlow/TensorShape.swift]: Our local convention is to omit `self.` unless it's necessary.

File: stdlib/public/TensorFlow/TensorShape.swift [hunk 2]
```diff
@@ -158,3 +160,13 @@ extension TensorShape : Codable {
...
+  }
+
+  public init?(_ pythonObject: PythonObject) {
+    self.init(Array<Int32>(pythonObject)!)
```
Comment (rxwei) [inline stdlib/public/TensorFlow/TensorShape.swift]: 1. Use short hand type names when possible for array and dictionary types.
2. This is a failable initializer that allows you to return `nil`. The use of `!` is unsafe.

```suggestion
    guard let array = [Int32](pythonObject) else {
      return nil
    }
    self.init(array)
```

**Review by rxwei — CHANGES_REQUESTED** [review]

**Comment by rxwei** [comment]
Have you tested something like `np.reshape(x, [1, 1])`?

**Comment by rxwei** [comment]
Thanks for tackling this!

**Comment by realdoug** [comment]
@rxwei in your example is ```x``` a TensorShape instance?  If so, that works as expected 👍 

I made the updates as well.

**Comment by rxwei** [comment]
‘x’ should be a numpy array.

File: test/Python/python_runtime.swift [hunk 3]
```diff
@@ -309,6 +318,11 @@ PythonRuntimeTestSuite.testWithLeakChecking("PythonConvertible") {
   expectEqual(five, UInt64(5).pythonObject)
   expectEqual(five, Float(5).pythonObject)
   expectEqual(five, Double(5).pythonObject)
+
+  expectEqual(intArray, Array([2, 3]).pythonObject)
+  expectEqual(dict, Dictionary<String, Int32>(["abc": 7]).pythonObject)
```
Comment (rxwei) [inline test/Python/python_runtime.swift]:
```suggestion
  expectEqual(dict, ["abc" : 7].pythonObject)
```

File: test/Python/python_runtime.swift [hunk 4]
```diff
@@ -309,6 +318,11 @@ PythonRuntimeTestSuite.testWithLeakChecking("PythonConvertible") {
   expectEqual(five, UInt64(5).pythonObject)
   expectEqual(five, Float(5).pythonObject)
   expectEqual(five, Double(5).pythonObject)
+
+  expectEqual(intArray, Array([2, 3]).pythonObject)
```
Comment (rxwei) [inline test/Python/python_runtime.swift]: Array literal `[2, 3]` already resolves to an `Array`.
```suggestion
  expectEqual(intArray, [2, 3].pythonObject)
```

File: test/Python/python_runtime.swift [hunk 5]
```diff
@@ -285,6 +288,10 @@ PythonRuntimeTestSuite.testWithLeakChecking("ConvertibleFromPython") {
 
   expectEqual("abc", String(string))
 
+  expectEqual([2, 3], Array(intArray))
+  expectEqual(TensorShape(2, 3), TensorShape(intArray))
+  expectEqual(["abc": 97], Dictionary<String, Int32>(dict))
```
Comment (rxwei) [inline test/Python/python_runtime.swift]: In the Swift core repo we put a space before `:`. This is a bit different from tensorflow/swift-apis.

```suggestion
  expectEqual(["abc" : 97], Dictionary<String, Int32>(dict))
```

File: test/Python/python_runtime.swift [hunk 6]
```diff
@@ -258,6 +259,8 @@ PythonRuntimeTestSuite.testWithLeakChecking("ConvertibleFromPython") {
   let five: PythonObject = 5
   let half: PythonObject = 0.5
   let string: PythonObject = "abc"
+  let intArray: PythonObject = [2, 3]
+  let dict: PythonObject = ["abc": 97]
```
Comment (rxwei) [inline test/Python/python_runtime.swift]:
```suggestion
  let dict: PythonObject = ["abc" : 97]
```

File: test/Python/python_runtime.swift [hunk 7]
```diff
@@ -293,6 +300,8 @@ PythonRuntimeTestSuite.testWithLeakChecking("ConvertibleFromPython") {
 PythonRuntimeTestSuite.testWithLeakChecking("PythonConvertible") {
   let minusOne: PythonObject = -1
   let five: PythonObject = 5
+  let intArray: PythonObject = [2, 3]
+  let dict: PythonObject = ["abc": 7]
```
Comment (rxwei) [inline test/Python/python_runtime.swift]:
```suggestion
  let dict: PythonObject = ["abc" : 7]
```

**Comment by rxwei** [comment]
Could you add numpy reshape tests to https://github.com/apple/swift/blob/tensorflow/test/Python/numpy_conversion.swift?

**Comment by realdoug** [comment]
@rxwei ok, I updated the syntax in all places in this file for consistency, verified that ```np.reshape``` works on the 0.2 release as well as this branch and added some tests for it.

File: test/Python/numpy_conversion.swift [hunk 8]
```diff
@@ -35,6 +35,17 @@ NumpyConversionTests.test("array-conversion") {
   let numpyArray2D = np.ones([2, 3])
   expectNil(Array<Float>(numpy: numpyArray2D))
 
+  let reshaped = np.reshape(numpyArray2D, 6)
+  if let array = expectNotNil(Array<Double>(numpy: reshaped)) {
+    expectEqual([1.0, 1.0, 1.0, 1.0, 1.0, 1.0], array)
+  }
+
+  let numpyArray1D = np.ones(28)
+  let reshaped3D = np.reshape(numpyArray1D, [2, 7, 2])
```
Comment (rxwei) [inline test/Python/numpy_conversion.swift]: Sorry, i meant to ask you to add reshape tests to https://github.com/apple/swift/blob/4e84b9267991c2e88342c83e8698914e335a9e44/test/TensorFlowRuntime/numpy_conversion.swift, where `TensorShape` is available.

File: test/TensorFlowRuntime/numpy_conversion.swift [hunk 9]
```diff
@@ -52,6 +52,12 @@ NumpyConversionTests.test("shaped-array-conversion") {
                 array)
   }
 
+  let reshaped = np.reshape(numpyArrayInt32, [2, 3])
```
Comment (rxwei) [inline test/TensorFlowRuntime/numpy_conversion.swift]:
```suggestion
  let reshaped = np.reshape(numpyArrayInt32, [2, 3] as TensorShape)
```
This is not using `TensorShape`, which kind of misses the point. Could you make the shape be an explicit `TensorShape`?

File: test/TensorFlowRuntime/numpy_conversion.swift [hunk 10]
```diff
@@ -95,6 +101,12 @@ NumpyConversionTests.test("tensor-conversion") {
                 tensor.array)
   }
 
+  let reshaped = np.reshape(numpyArrayInt32, [2, 3])
```
Comment (rxwei) [inline test/TensorFlowRuntime/numpy_conversion.swift]: Same here.

File: test/TensorFlowRuntime/numpy_conversion.swift (hunk 9 above)
Comment (realdoug) [reply test/TensorFlowRuntime/numpy_conversion.swift]: 🤦‍♂️ ok i see what you were saying now.  How is this look? @rxwei

File: test/Python/python_runtime.swift [hunk 11]
```diff
@@ -5,6 +5,7 @@
 
 import Python
 import StdlibUnittest
+import struct TensorFlow.TensorShape
```
Comment (rxwei) [inline test/Python/python_runtime.swift]: These tests should also move to `TensorFlowRuntime/numpy_conversion.swift`. Python tests should not depend on the `TensorFlow` module.

File: test/TensorFlowRuntime/numpy_conversion.swift (hunk 9 above)
Comment (rxwei) [reply test/TensorFlowRuntime/numpy_conversion.swift]: Looks great.

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Review by rxwei — APPROVED** [review]

**Comment by rxwei** [comment]
@swift-ci please clean test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by realdoug** [comment]
@rxwei I believe this is a real build failure.  I was able to repro on a fresh Linux install after pasting my TensorShape extension onto HEAD.  Looking into it but not sure f you have any gut thoughts based on the error message.

**Comment by realdoug** [comment]
@rxwei Can you rerun CI?

1) I added logic to detect ```len()``` and handle the case where ```pythonObject``` is a single int.  
2) There is also a [bug](https://bugs.swift.org/browse/TF-426) that I was able to reproduce without any python related code & changing ```TensorShape(1)``` to ```TensorShape([1])``` steps around it for now.

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
ugh...

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

File: stdlib/public/TensorFlow/TensorShape.swift [hunk 12]
```diff
@@ -10,6 +10,8 @@
 //
 //===----------------------------------------------------------------------===//
 
+import Python
```
Comment (dan-zheng) [inline stdlib/public/TensorFlow/TensorShape.swift]: Could you please do the following?
* Rename `stdlib/public/TensorFlow/NumpyConversion.swift` to `PythonConversion.swift`.
  * Make appropriate changes to the file header so it's not NumPy-specific.
* Move `import Python` and the `PythonConvertible` logic from this file to `PythonConversion.swift`.

That would be much appreciated! 🙂

Comment (realdoug) [reply stdlib/public/TensorFlow/TensorShape.swift]: @dan-zheng sure thing: https://github.com/apple/swift/pull/23891

**Comment by rxwei** [comment]
Hi @realdoug, since this PR changed the semantics of `TensorShape(0)` to rely on Python, this may have broken some tests in other platforms. I'm going to revert this PR since we are approaching our 0.3 release. Let me know when you have a PR that splits `PythonConvertible` into two protocols!

**Pull request merged**